pilvytis/order_status_tracker.go

```diff
@@ -0,0 +1,128 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package pilvytis
+
+import (
+	"sync"
+	"time"
+
+	"github.com/mysteriumnetwork/node/eventbus"
+	"github.com/mysteriumnetwork/node/identity"
+	"github.com/rs/zerolog/log"
+)
+
+type OrderSummary struct {
+	ID              uint64
+	IdentityAddress string
+	Status          OrderStatus
+}
+
+type orderProvider interface {
+	GetPaymentOrders(id identity.Identity) ([]OrderResponse, error)
+}
+
+type StatusTracker struct {
+	api              orderProvider
+	identityProvider identityProvider
+	eventBus         eventbus.Publisher
+	orders           []OrderSummary
+
+	pauseCondition func(response []OrderResponse) bool
+	updateInterval time.Duration
+	startOnce      sync.Once
+	stopCh         chan struct{}
+	stopOnce       sync.Once
+}
+
+func NewStatusTracker(api orderProvider, identityProvider identityProvider, eventBus eventbus.Publisher, updateInterval time.Duration) *StatusTracker {
+	return &StatusTracker{
+		api:              api,
+		identityProvider: identityProvider,
+		eventBus:         eventBus,
+		orders:           []OrderSummary{},
+		updateInterval:   updateInterval,
+		stopCh:           make(chan struct{}),
+	}
+}
+
+func (t *StatusTracker) getOrCreate(id uint64, identityAddress string) *OrderSummary {
+	for i := range t.orders {
+		if t.orders[i].ID == id {
+			return &t.orders[i]
+		}
+	}
+	newEntry := OrderSummary{ID: id, IdentityAddress: identityAddress}
+	t.orders = append(t.orders, newEntry)
+	return &newEntry
+}
+
+func (t *StatusTracker) Track() {
+	trackingFunc := func() {
+		log.Info().Msg("Tracking order statuses...")
+		t.stopOnce = sync.Once{}
+
+		updateTicker := time.NewTicker(t.updateInterval)
+		for {
+			select {
+			case <-t.stopCh:
+				return
+			case <-updateTicker.C:
+				err := t.update()
+				if err != nil {
+					log.Err(err).Msg("Could not update order statuses")
+				}
+			}
+		}
+	}
+	t.startOnce.Do(func() {
+		go trackingFunc()
+	})
+}
+
+func (t *StatusTracker) Pause() {
+	t.stopOnce.Do(func() {
+		log.Info().Msg("Pausing order status tracking")
+		t.stopCh <- struct{}{}
+		t.startOnce = sync.Once{}
+	})
+}
+
+func (t *StatusTracker) update() error {
+	log.Info().Msg("Updating order statuses")
+	keepTracking := false
+	for _, id := range t.identityProvider.GetIdentities() {
+		newOrders, err := t.api.GetPaymentOrders(id)
+		if err != nil {
+			return err
+		}
+		for _, newOrder := range newOrders {
+			order := t.getOrCreate(newOrder.ID, newOrder.Identity)
+			if order.Status != newOrder.Status {
+				order.Status = newOrder.Status
+				t.eventBus.Publish(AppTopicOrderStatusChanged, AppEventOrderStatusChanged{*order})
+			}
+			if newOrder.Status.Incomplete() {
```

Comment:
I think if `GetPaymentOrders` returns orders in unexpected order this will make things fail no?
If for example we get `Pending, Pending, Complete` this will stop. Or do we not care about that?

Reply:
It will not stop if any of the orders are incomplete. The order does not matter here.

Reply:
Actually, not reversing, but added `keepTracking = true` to the error case branch.